File: practica/C++/5.2.cpp
#include <iostream>
#include <string>

#include <windows.h>
#include <conio.h>

using namespace std;

//!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

//генерування матриці
void FuncGenerateArray
(
    int CounterArray, //лічильник масиву
    int Array[10][10] //масив
);

//вивід матриці
void FuncOutputArray
(
    int CounterArray, //лічильник масиву
    int Array[10][10], //масив
    string Message //повідомлення
);

//копіювання масиву
void FuncCopyArray
(
    int CounterArray, //лічильник масиву
    int Array[10][10], //масив
    int CopyArray[10 * 10] //масив копія
);

//вивід копіювального масиву
void FuncOutputCopyArray
(
    int CounterArray, //лічильник масиву
    int CopyArray[10 * 10], //масив копія
    string Message //повідомлення
);

//сортування масиву методом вибору
void FuncSortArray
(
    int CounterArray, //лічильник масиву
    int CopyArray[10 * 10] //масив копія
);

//функція створення потрібної матриці зі завдання
void FuncCreateTrueMatrix
(
    int CounterArray, //лічильник масиву
    int Array[10][10], //масив
    int CopyArray[10 * 10] //масив копія
);

//|||||||||||||||||||||||||||||||||||||||||||||||||||||
//|||||||||||||||||||||||||||||||||||||||||||||||||||||
//|||||||||||||||||||||||||||||||||||||||||||||||||||||

int main()
{
    SetConsoleCP(1251);
    SetConsoleOutputCP(1251);

    int CounterArray; //лічильник масиву

    while(true) //цикл вводу розмірності матриці
    {
        cout << "Введіть розмірність матриці: ";
        cin >> CounterArray;
        
        if( (CounterArray <= 10) && (CounterArray > 1) )
        {
            break;
        }
        else
        {
            cout << "\t" << "Розмірність повинна бути в діпазоні [2; 10]" << "\n\n\n\n";
        }
    }

    int Array[10][10]; //головний масив
    int CopyArray[10 * 10]; //одномірний масив з усіма елементами

    FuncGenerateArray(CounterArray, Array);
    FuncOutputArray(CounterArray, Array, "Вивід матриці після генерування");
    FuncCopyArray(CounterArray, Array, CopyArray);
    FuncOutputCopyArray(CounterArray, CopyArray, "Вивід масиву зі значеннями матриці");
    FuncSortArray(CounterArray, CopyArray);
    FuncOutputCopyArray(CounterArray, CopyArray, "Вивід масиву зі значеннями матриці у сортувальному вигляді");
    FuncCreateTrueMatrix(CounterArray, Array, CopyArray);
    FuncOutputArray(CounterArray, Array, "Вивід матриці у завданні");

    getch();

    return 0;
}

//|||||||||||||||||||||||||||||||||||||||||||||||||||||
//|||||||||||||||||||||||||||||||||||||||||||||||||||||
//|||||||||||||||||||||||||||||||||||||||||||||||||||||

//генерування матриці
void FuncGenerateArray
(
    int CounterArray, //лічильник масиву
    int Array[10][10] //масив
)
{
    cout << "Заповнення масиву випадковими числами:" << "\n" << "\t";
    
    for(int i = 0; i < CounterArray; i = i + 1)
    {
        for(int j = 0; j < CounterArray; j = j + 1)
        {
            Array[i][j] = rand() % 9;

            cout << "елемент №" << i + 1 << "-" << j + 1 << ": " << Array[i][j] << "\n" << "\t";
        }
    }

    cout << "\n\n\n";
}

//вивід матриці
void FuncOutputArray
(
    int CounterArray, //лічильник масиву
    int Array[10][10], //масив
    string Message //повідомлення
)
{
    cout << Message << ":" << "\n";
    
    for(int i = 0; i < CounterArray; i = i + 1)
    {
        cout << "\t";
        
        for(int j = 0; j < CounterArray; j = j + 1)
        {
            cout << Array[i][j] << " ";
        }

        cout << "\n";
    }

    cout << "\n\n\n";
}

//копіювання масиву
void FuncCopyArray
(
    int CounterArray, //лічильник масиву
    int Array[10][10], //масив
    int CopyArray[10 * 10] //масив копія
)
{
    int CounterCopyArray = 0; //лічильник масиву копії

    for(int i = 0; i < CounterArray; i = i + 1)
    {
        for(int j = 0; j < CounterArray; j = j + 1)
        {
            CopyArray[CounterCopyArray] = Array[i][j];

            CounterCopyArray = CounterCopyArray + 1;
        }
    }
}

//вивід копіювального масиву
void FuncOutputCopyArray
(
    int CounterArray, //лічильник масиву
    int CopyArray[10 * 10], //масив копія
    string Message //повідомлення
)
{
    cout << Message << ":" << "\n" << "\t";

    for(int i = 0; i < (CounterArray * CounterArray); i = i + 1)
    {
        if(((i + 1) % 25 ) == 0)
        {
            cout << CopyArray[i] << "\n" << "\t";
        }
        else
        {
            cout << CopyArray[i] << " ";
        }
    }

    cout << "\n\n\n";
}

//сортування масиву методом вибору
void FuncSortArray
(
    int CounterArray, //лічильник масиву
    int CopyArray[10 * 10] //масив копія
)
{
    for(int i = 0; i < (CounterArray * CounterArray - 1); i = i + 1)
    {
        for(int j = i + 1; j < (CounterArray * CounterArray); j = j + 1)
        {
            if(CopyArray[i] > CopyArray[j])
            {
                int TempVar = CopyArray[i];
                CopyArray[i] = CopyArray[j];
                CopyArray[j] = TempVar;
            }
        }
    }
}

//функція створення потрібної матриці зі завдання
void FuncCreateTrueMatrix
(
    int CounterArray, //лічильник масиву
    int Array[10][10], //масив
    int CopyArray[10 * 10] //масив копія
)
{
    int CounterCopyArray = 0; //лічильник масиву копії

    for(int i = 0; i < CounterArray; i = i + 1)
    {
        if(((i + 1) % 2) != 0) //непарний рядок буде за спаданням
        {
            for(int j = CounterArray - 1; j >= 0; j = j - 1)
            {
                Array[i][j] = CopyArray[CounterCopyArray];

                CounterCopyArray = CounterCopyArray + 1;
            }
        }
        else //парний рядок буде за зростанням
        {
            for(int j = 0; j < CounterArray; j = j + 1)
            {
                Array[i][j] = CopyArray[CounterCopyArray];

                CounterCopyArray = CounterCopyArray + 1;
            }
        }
    }
}

File: practica/python/5.2.py
from msvcrt import getch
from random import randint

CounterArray = 0 #лічильник масиву
Array = []
CopyArray = []

#генерування масиву
def FuncGenerateArray(): 
    print("Заповнення масиву випадковими числами:", sep = "", end = "\n\t")
    for i in range(0, CounterArray, 1):
        TempArray = []

        for j in range(0, CounterArray, 1):
            TempArray.append(randint(0, 9))
            print("елемент №", i + 1, "-", j + 1, ": ", TempArray[j], sep = "", end = "\n\t")

        Array.append(TempArray)

    print(end = "\n\n\n")

#вивід масиву
def FuncOutputArray(Message):
    print(Message, ":", sep = "")
    for i in range(0, len(Array), 1):
        print(end = "\t")

        for j in range(0, len(Array), 1):
            print(Array[i][j], end = " ")

        print()

    print(end = "\n\n\n")

#копіювання масиву
def FuncCopyArray():
    for i in range(0, len(Array), 1):
        for j in range(0, len(Array), 1):
            CopyArray.append(Array[i][j])

#вивід копіювального масиву
def FuncOutputCopyArray(Message):
    print(Message, ":", sep = "", end = "\n\t")

    for i in range(0, len(CopyArray), 1):
        if(((i + 1) % 25 ) == 0):
            print(CopyArray[i], end = "\n\t")
        else:
            print(CopyArray[i], end = " ")

    print(end = "\n\n\n")

#сортування масиву методом вибору
def FuncSortArray():
    for i in range(0, len(CopyArray) - 1, 1):
        for j in range(i + 1, len(CopyArray), 1):
            if(CopyArray[i] > CopyArray[j]):
                TempVar = CopyArray[i]
                CopyArray[i] = CopyArray[j]
                CopyArray[j] = TempVar



#функція створення потрібної матриці зі завдання
def FuncCreateTrueMatrix():
    CounterCopyArray = 0; #лічильник масиву копії

    for i in range(0, len(Array), 1):
        #непарний рядок буде за спаданням
        if(((i + 1) % 2) != 0):
            for j in range(len(Array) - 1, -1, -1):
                Array[i][j] = CopyArray[CounterCopyArray]
                CounterCopyArray = CounterCopyArray + 1
        #парний рядок буде за зростанням
        else:
            for j in range(0, len(Array), 1):
                Array[i][j] = CopyArray[CounterCopyArray]
                CounterCopyArray = CounterCopyArray + 1

#ГОЛОВНА ПРОГРАМА
#цикл вводу розмірності матриці
while(True):
    CounterArray = int(input("Введіть розмірність матриці: "))
    
    if( (CounterArray <= 10) and (CounterArray > 1) ):
        break
    else:
        print("\t Розмірність повинна бути в діпазоні [2 10]", end = "\n\n")

FuncGenerateArray()
FuncOutputArray("Вивід матриці до сортування")
FuncCopyArray()
FuncOutputCopyArray("Вивід масиву значень матриці")
FuncSortArray()
FuncOutputCopyArray("Вивід масиву значень матриці у сортованому вигляді")
FuncCreateTrueMatrix()
FuncOutputArray("Вивід матриці з сортування для завдання")

getch()